### PR TITLE
Improvements for virsh chapter of virtualization guide

### PR DESCRIPTION
**Pull request by jfehlig**
### Description
This PR improves some existing text in the virsh chapter of the virtualization guide and adds some new text for new settings.

### Checklist
* Check all items that apply.

*Are backports required?*

ONLY patch 1 needs to be backported to the following 

- [x ] To maintenance/SLE15SP1
- [x ] To maintenance/SLE15SP0
- [ x] To maintenance/SLE12SP5

Patch 2 and 3 are new for SLE15 SP2.

### CODE REVIEW

File: xml/libvirt_configuration_virsh.xml [hunk 1]
```diff
@@ -224,6 +162,39 @@ Product Name: Standard PC (Q35 + ICH9, 2009)</screen>
...
+ <sect1 xml:id="sec-libvirt-hypervisor-features-virsh">
+  <title>Configuring Hypervisor Features</title>
+
+  <para>
+   libvirt automatically enables a default set of hypervisor features that
+   are sufficient in most circumstances, but also allows enabling and disabling
+   features as needed. As an example, Xen does not support enabling PCI
+   passthrough by default. It must be enabled with the
+   <literal>passthrough</literal> setting. Hypervisor features can be configured
+   with &virsh;. Look for the <literal>&lt;features</literal> element in the
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:174]:
```suggestion
   with &virsh;. Look for the <literal>&lt;features&gt;</literal> element in the
```

File: xml/libvirt_configuration_virsh.xml [hunk 2]
```diff
@@ -224,6 +162,39 @@ Product Name: Standard PC (Q35 + ICH9, 2009)</screen>
...
+   &vmguest;'s configuration file and adjust the various featues as required.
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:175]:
```suggestion
   &vmguest;'s configuration file and adjust the various features as required.
```

File: xml/libvirt_configuration_virsh.xml [hunk 3]
```diff
@@ -224,6 +162,39 @@ Product Name: Standard PC (Q35 + ICH9, 2009)</screen>
    </para>
   </tip>
  </sect1>
+ <sect1 xml:id="sec-libvirt-hypervisor-features-virsh">
+  <title>Configuring Hypervisor Features</title>
+
+  <para>
+   libvirt automatically enables a default set of hypervisor features that
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:169]:
```suggestion
   <command>libvirt</command> automatically enables a default set of hypervisor features that
```

File: xml/libvirt_configuration_virsh.xml [hunk 4]
```diff
@@ -1273,6 +1244,93 @@ pci_0000_04_11_5</screen>
...
  </sect1>
+ <sect1 xml:id="sec-libvirt-config-controllers-virsh">
+  <title>Configuring Controller Devices</title>
+
+  <para>
+   <command>libvirt</command> generally manages controllers automatically
+   based on the type of virtual devices used by the &vmguest;. If the &vmguest;
+   contains PCI and SCSI devices, PCI and SCSI controllers will be created and
+   managed automatically. <command>libvirt</command> will also model controllers
+   that are hypervisor-specifiic, for example a virtio-serial controller for KVM
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1255]:
```suggestion
   that are hypervisor-specific, for example a <literal>virtio-serial</literal> controller for KVM
```

File: xml/libvirt_configuration_virsh.xml [hunk 5]
```diff
@@ -1273,6 +1244,93 @@ pci_0000_04_11_5</screen>
...
+   &vmguest;s or a xenbus controller for Xen &vmguest;s. Although the default
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1256]:
```suggestion
   &vmguest;s or a <literal>xenbus</literal> controller for Xen &vmguest;s. Although the default
```

File: xml/libvirt_configuration_virsh.xml [hunk 6]
```diff
@@ -1273,6 +1244,93 @@ pci_0000_04_11_5</screen>
...
+   controllers and their configuration are generally fine there may be use cases
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1257]:
```suggestion
   controllers and their configuration are generally fine, there may be use cases
```

File: xml/libvirt_configuration_virsh.xml [hunk 7]
```diff
@@ -1273,6 +1244,93 @@ pci_0000_04_11_5</screen>
...
+   where controllers or their attributes need to be adjusted manually. For example
+   a virtio-serial controller may need more ports or a xenbus controller may need
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1259]:
```suggestion
   a virtio-serial controller may need more ports, or a xenbus controller may need
```

File: xml/libvirt_configuration_virsh.xml [hunk 8]
```diff
@@ -1273,6 +1244,93 @@ pci_0000_04_11_5</screen>
...
+   more memory or more virtual interrupts.
+  </para> 
+
+  <para>
+   The xenbus controller is a bit unique in that it serves as the controller for
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1264]:
```suggestion
   The xenbus controller is unique, in that it serves as the controller for
```

File: xml/libvirt_configuration_virsh.xml [hunk 9]
```diff
@@ -1273,6 +1244,93 @@ pci_0000_04_11_5</screen>
...
+
+ &prompt.sudo;xen-diag gnttab_query_size 0
+domid=0: nr_frames=1, max_nr_frames=256
+
+&prompt.sudo;xen-diag gnttab_query_size 3
+domid=3: nr_frames=3, max_nr_frames=32
+</screen>
+
+  <para>
+   The sle15sp1 guest is using only 3 frames out of 32. If you are seeing
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1294]:
```suggestion
   The <literal>sle15sp1</literal> guest is using only three frames out of 32. If you are seeing
```

File: xml/libvirt_configuration_virsh.xml [hunk 10]
```diff
@@ -1273,6 +1244,93 @@ pci_0000_04_11_5</screen>
...
+   performance issues, and log entries that point to insufficient frames,
+   increase the value with &virsh;. Look for the <literal>&lt;controller
+   type='xenbus'</literal> line in the guest's configuration file and add the
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1297]:
```suggestion
   type='xenbus'&gt;</literal> line in the guest's configuration file, and add the
```

File: xml/libvirt_configuration_virsh.xml [hunk 11]
```diff
@@ -1273,6 +1244,93 @@ pci_0000_04_11_5</screen>
...
+   Save your changes and restart the guest. Now it should show your change:
+  </para>
+
+<screen>&prompt.sudo;xen-diag gnttab_query_size 3
+domid=3: nr_frames=3, max_nr_frames=40
+</screen>
+
+  <para>
+   Similar to maxGrantFrames, the xenbus controller also supports
+   maxEventChannels. Event channels are like paravirtual interrupts, and in
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1315]:
```suggestion
   <option>maxEventChannels</option>. Event channels are like paravirtual interrupts, and in
```

File: xml/libvirt_configuration_virsh.xml [hunk 12]
```diff
@@ -1273,6 +1244,93 @@ pci_0000_04_11_5</screen>
...
+   conjuction with grant frames form a data transfer mechanism for paravirtual
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1316]:
```suggestion
   conjuction with grant frames, form a data transfer mechanism for paravirtual
```

**Comment by jfehlig** [comment]
Hi Liam,

I noticed another typo while reviewing your review :-). It was in a line touched by one of your suggestions. After accepting all of your suggestions in a batch commit, I amended the commit introducing the typo, fixed up your suggestions commit while rebasing, the force pushed the mess. I'm a little green on github's merge request workflow so hopefully I didn't screw it up too badly.

File: xml/libvirt_configuration_virsh.xml [hunk 13]
```diff
@@ -1273,6 +1244,93 @@ pci_0000_04_11_5</screen>
...
+   <literal>max_grant_frames</literal> attribute sets how many grant frames, or
+   blocks of shared memory, are allocated to the <literal>xenbus</literal>
+   controller for each &vmguest;.
+  </para>
+
+  <para>
+   The default of 32 is enough in most circumstances, but a &vmguest; with a large
+   number of I/O devices and an I/O intensive workload may experience performance
+   issues due to grant frame exhaustion. The <command>xen-diag</command> can be
+   used to check the current used and maximum <literal>max_grant_frames</literal>
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1276]:
```suggestion
   used to check the current and maximum <literal>max_grant_frames</literal>
```

File: xml/libvirt_configuration_virsh.xml [hunk 14]
```diff
@@ -1273,6 +1244,93 @@ pci_0000_04_11_5</screen>
...
+   all Xen paravirtual devices. If a &vmguest; has many disk and/or network
+   devices the controller may need more memory. Xen's
+   <literal>max_grant_frames</literal> attribute sets how many grant frames, or
+   blocks of shared memory, are allocated to the <literal>xenbus</literal>
+   controller for each &vmguest;.
+  </para>
+
+  <para>
+   The default of 32 is enough in most circumstances, but a &vmguest; with a large
+   number of I/O devices and an I/O intensive workload may experience performance
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1274]:
```suggestion
   number of I/O devices and an I/O-intensive workload may experience performance
```

File: xml/libvirt_configuration_virsh.xml [hunk 15]
```diff
@@ -224,6 +162,39 @@ Product Name: Standard PC (Q35 + ICH9, 2009)</screen>
...
+ <sect1 xml:id="sec-libvirt-hypervisor-features-virsh">
+  <title>Configuring Hypervisor Features</title>
+
+  <para>
+   <command>libvirt</command> automatically enables a default set of hypervisor features that
+   are sufficient in most circumstances, but also allows enabling and disabling
+   features as needed. As an example, Xen does not support enabling PCI
+   passthrough by default. It must be enabled with the
+   <literal>passthrough</literal> setting. Hypervisor features can be configured
+   with &virsh;. Look for the <literal>&lt;features&gt;</literal> element in the
```
Comment (taroth21) [inline xml/libvirt_configuration_virsh.xml:174]:
```suggestion
   with &virsh;. Look for the <tag>&lt;features&gt;</tag> element in the
```

File: xml/libvirt_configuration_virsh.xml [hunk 16]
```diff
@@ -1273,6 +1244,93 @@ pci_0000_04_11_5</screen>
...
+
+  <para>
+   <command>libvirt</command> generally manages controllers automatically
+   based on the type of virtual devices used by the &vmguest;. If the &vmguest;
+   contains PCI and SCSI devices, PCI and SCSI controllers will be created and
+   managed automatically. <command>libvirt</command> will also model controllers
+   that are hypervisor-specific, for example a <literal>virtio-serial</literal> controller for KVM
+   &vmguest;s or a <literal>xenbus</literal> controller for Xen &vmguest;s. Although the default
+   controllers and their configuration are generally fine, there may be use cases
+   where controllers or their attributes need to be adjusted manually. For example
```
Comment (taroth21) [inline xml/libvirt_configuration_virsh.xml:1258]: Add a comma? ```For example,```

Comment (lproven) [reply xml/libvirt_configuration_virsh.xml:1258]: Good idea. Done.

**Comment by jfehlig** [comment]
It's not clear to me what is needed to merge this request. @taroth21 are you waiting for something from me?

**Pull request merged**